x-pack/plugins/lens/public/indexpattern_datasource/operations/layer_helpers.ts

```diff
@@ -342,3 +463,116 @@ export function updateLayerIndexPattern(
+          errors.push(
+            i18n.translate('xpack.lens.indexPattern.missingReferenceError', {
+              defaultMessage: 'Dimension {dimensionLabel} is incomplete',
+              values: {
+                // @ts-expect-error references are not statically analyzed yet
+                dimensionLabel: column.label,
+              },
+            })
+          );
+        } else {
+          const referenceColumn = layer.columns[referenceId]!;
+          const requirements =
+            // @ts-expect-error not statically analyzed
+            operationDefinitionMap[column.operationType].requiredReferences[index];
+          const isValid = isColumnValidAsReference({
+            validation: requirements,
+            column: referenceColumn,
+          });
+
+          if (!isValid) {
+            errors.push(
+              i18n.translate('xpack.lens.indexPattern.invalidReferenceConfiguration', {
+                defaultMessage: 'Dimension {dimensionLabel} does not have a valid configuration',
+                values: {
+                  // @ts-expect-error references are not statically analyzed yet
+                  dimensionLabel: column.label,
+                },
+              })
+            );
+          }
+        }
+      });
+    }
+  });
+
+  return errors.length ? errors : undefined;
+}
+
+export function isReferenced(layer: IndexPatternLayer, columnId: string): boolean {
+  const allReferences = Object.values(layer.columns).flatMap((col) =>
+    'references' in col
+      ? // @ts-expect-error not statically analyzed
+        col.references
+      : []
+  );
+  return allReferences.includes(columnId);
+}
+
+function isColumnValidAsReference({
+  column,
+  validation,
+}: {
+  column: IndexPatternColumn;
+  validation: RequiredReference;
+}): boolean {
+  if (!column) return false;
+  const operationType = column.operationType;
+  const operationDefinition = operationDefinitionMap[operationType];
+  return (
+    validation.input.includes(operationDefinition.input) &&
+    (!validation.specificOperations || validation.specificOperations.includes(operationType)) &&
+    validation.validateMetadata(column)
+  );
+}
+
+function isOperationAllowedAsReference({
+  operationType,
+  validation,
+  field,
+}: {
+  operationType: OperationType;
+  validation: RequiredReference;
+  field?: IndexPatternField;
+}): boolean {
+  const operationDefinition = operationDefinitionMap[operationType];
+
+  let hasValidMetadata = true;
+  if (field && operationDefinition.input === 'field') {
+    const metadata = operationDefinition.getPossibleOperationForField(field);
+    hasValidMetadata = Boolean(metadata) && validation.validateMetadata(metadata!);
+  } else if (operationDefinition.input !== 'field') {
+    const metadata = operationDefinition.getPossibleOperation();
+    hasValidMetadata = Boolean(metadata) && validation.validateMetadata(metadata!);
+  } else {
+    // TODO: How can we validate the metadata without a specific field?
```

Comment:
That's a very good question and we have to answer it in some way for the UI so we can provide a list of sub-functions. One option would be to go through all fields of the current index pattern and return true if one of them matches.

Reply:
Exactly, this can't be done purely in the code here.